README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# tidyestimate <img src='man/figures/logo.png' align="right" height="138" />

<!-- badges: start -->

[![R-CMD-check](https://github.com/KaiAragaki/tidyestimate/workflows/R-CMD-check/badge.svg)](https://github.com/KaiAragaki/tidyestimate/actions)
<!-- badges: end -->

The ESTIMATE package has been fundamental for inferring tumor purity
from expression data, but its documentation is lacking, and its
functions sometimes overstep their bounds while not doing enough. This
package is a refresh of ESTIMATE with the goal of maintaining the
excellent backbone of the package while increasing its documentation and
function scope.

## Installation

You can install the released version of tidyestimate from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("tidyestimate")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("KaiAragaki/tidy_estimate")
```

## Features

|              |               tidyestimate               |  ESTIMATE   |
|-------------:|:----------------------------------------:|:-----------:|
|        input | `data.frame`<br />`tibble`<br />`matrix` | `.GCT` file |
|       output |               `data.frame`               | `.GCT` file |
| `%>%`/`\|>`? |                    ✔️                    |     ✖️      |
|         size |                 &lt;1MB                  |    \~7MB    |

Additionally:

⚡ Faster. `tidyestimate` doesn’t do any file conversion.

📝 Better documentation. Functions are more clear about input
requirements and returns.

🕊️ Lighter. Less code, more readable (less to break, easier to fix).

💪 Robust. `tidyestimate` does conservative alias matching to allow
compatibility with both old and new gene identifiers.

## Quickstart

Evaluating tumor purity with `tidyestimate` is simple. `tidyestimate`
can take a `matrix` or `data.frame` (and thus a `tibble`). In this
example, we’ll be using the `ov` dataset, which is derived from the
original `estimate` package. It’s a matrix with expression data
(profiled using an array-based Affymetrix method) for 10 ovarian cancer
tumors.

``` r
library(tidyestimate)
```

``` r
dim(ov)
#> [1] 17256    10
```

``` r
head(ov)[,1:5]
#>             s516    s518    s519    s520    s521
#> C9orf152  4.8815  4.5757  3.7395  3.6960  4.1597
#> ELMO2     7.2981  7.5554  7.5332  7.3824  7.3079
#> CREB3L1   5.5692  5.7004  5.9597  5.7700  5.2190
#> RPS11    13.3899 13.8488 13.6429 13.6546 13.5698
#> PNMA1     9.3480 10.0092 10.4310  9.5399  9.6423
#> MMP2      7.6182  8.0369  8.9551 10.3875  7.4141
```

Tumor purity can be predicted like so:

``` r
scores <- ov |> 
  filter_common_genes(id = "hgnc_symbol", tell_missing = FALSE, find_alias = TRUE) |> 
  estimate_score(is_affymetrix = TRUE)
#> Joining, by = "hgnc_symbol"
#> Joining, by = "hgnc_symbol"
#> 461 of 488 missing genes found matches using aliases.
#> Joining, by = "hgnc_symbol"
#> 
#> Found 10364 of 10391 genes (99.74%) in your dataset.
#> Number of stromal_signature genes in data: 141 (out of 141)
#> Number of immune_signature genes in data: 141 (out of 141)
scores
#>    sample    stromal     immune   estimate    purity
#> 1    s516 -285.49841  165.75062  -119.7478 0.8323791
#> 2    s518 -429.16931   99.71302  -329.4563 0.8490421
#> 3    s519  -60.98619 -368.70314  -429.6893 0.8567232
#> 4    s520 1927.51431 2326.15984  4253.6742 0.3348246
#> 5    s521 -673.84954  141.72775  -532.1218 0.8643812
#> 6    s522 1447.95517 1166.51854  2614.4737 0.5497248
#> 7    s523 -271.15756 -928.44921 -1199.6068 0.9094242
#> 8    s525  965.61804 1310.27775  2275.8958 0.5905450
#> 9    s526  545.99467 2149.10473  2695.0994 0.5398002
#> 10   s527 -710.44370 1303.08009   592.6364 0.7699846
```

They can also be plotted in context of the Affymetrix profiled tumors
used to generate the ESTIMATE model:

``` r
scores |> 
  plot_purity(is_affymetrix = TRUE)
```

<img src="man/figures/README-unnamed-chunk-3-1.png" width="100%" /> A
more detailed version of this example can be found in the vignette of
this package.


NEWS.md
# tidyestimate 1.1.0

## Bugfixes

* When using `filter_common_genes`, supplying a dataframe that has a gene column that is NOT named `hgnc_symbol` or `entrezgene_id` now works as expected
* When using `filter_common_genes` and supplying entrez ids, the returned dataframe is now HGNC symbols to better flow into `estimate_score`
* Failure to supply `is_affymetrix` results in an earlier error, before calculating scores.

# tidyestimate 1.0.4

* CRAN release

# tidyestimate 1.0.3

# tidyestimate 1.0.2

* Added a `NEWS.md` file to track changes to the package.


R/plot_purity.R
#' Plot Affymetrix purity scores against ESTIMATE study purity scores
#'
#' @param scores a \code{data.frame}, usually one output from \code{estimate_score}
#' @param is_affymetrix logical. Are these data from an Affymetrix experiment?
#'   Must be \code{TRUE} - this is essentially a verification from the user
#'
#' @return a \code{ggplot}
#' @export
#'
#' @importFrom ggplot2 aes
#' @importFrom rlang .data
#' @examples
#' filter_common_genes(ov, id = "hgnc_symbol", tidy = FALSE, tell_missing = TRUE, find_alias = TRUE) |> 
#'   estimate_score(is_affymetrix = TRUE) |>
#'   plot_purity(is_affymetrix = TRUE)
#' 
plot_purity <- function(scores, is_affymetrix) {
        if(!is_affymetrix){
                stop("Non-Affymetrix data is not supported")
        }
        
        scores_nona <- dplyr::filter(scores, !is.na(.data$purity))
        
        global_min <- min(scores_nona$estimate, tidyestimate::purity_data_affy$estimate, na.rm = T)
        global_max <- max(scores_nona$estimate, tidyestimate::purity_data_affy$estimate, na.rm = T)
        purity_line <- dplyr::tibble(estimate = seq(from = global_min, 
                                                    to = global_max, 
                                                    by = 10),
                                     purity = cos(0.6049872018 + 0.0001467884 * .data$estimate))
        
        ggplot2::ggplot(tidyestimate::purity_data_affy, aes(x = .data$estimate)) + 
                ggplot2::geom_point(aes(y = .data$purity_observed), alpha = 0.2, shape = 1) +
                ggplot2::geom_line(aes(y = .data$ci_95_low), linetype = "dashed", alpha = 0.3) +
                ggplot2::geom_line(aes(y = .data$ci_95_high), linetype = "dashed", alpha = 0.3) +
                ggplot2::geom_line(data = purity_line, aes(y = .data$purity), alpha = 0.4) +
                ggplot2::geom_point(data = scores, aes(y = .data$purity), color = "red", size = 2, shape = 1) +
                ggrepel::geom_text_repel(data = scores, aes(y = .data$purity, label = .data$sample)) + 
                ggplot2::coord_cartesian(ylim = c(0, 1)) +
                ggplot2::theme(panel.background = ggplot2::element_rect(fill = NA, 
                                                                        color = "black"), 
                               panel.grid = ggplot2::element_blank()) +
                ggplot2::labs(x = "ESTIMATE score", y = "Tumor purity")
}


R/estimate_score.R
#' Infer tumor purity using the ESTIMATE algorithm
#'
#' Infer tumor purity by using single-sample gene-set-enrichment-analysis with
#' stromal and immune cell signatures.
#'
#' @param df a \code{data.frame} of expression data, where columns are tumors
#'   and rows are genes. Gene names must be in the first column, and in the form
#'   of HGNC symbols.
#' @param is_affymetrix logical. Is the expression data from an Affymetrix
#'   array?
#'
#' @details
#'
#' ESTIMATE (and this tidy implementation) infers tumor infiltration using two
#' gene sets: a stromal signature, and an immune signature (see
#' \code{tidyestimate::gene_sets}).
#'
#' Enrichment scores for each sample are calculated using an implementation of
#' single sample Gene Set Enrichment Analysis (ssGSEA). Briefly, expression is
#' ranked on a per-sample basis, and the density and distribution of gene
#' signature 'hits' is determined. An enrichment of hits at the top of the
#' expression ranking confers a positive score, while an enrichment of hits at
#' the bottom of the expression ranking confers a negative score.
#'
#' An 'ESTIMATE' score is calculated by adding the stromal and immune scores
#' together.
#'
#' For Affymetrix arrays, an equation to convert an ESTIMATE score to a
#' prediction of tumor purity has been developed by Yoshihara et al. (see
#' references). It takes the approximate form of:
#'
#' \deqn{purity = cos(0.61 + 0.00015 * ESTIMATE)}
#'
#' Values have been rounded to two significant figures for display purposes.
#'
#'
#' @return A \code{data.frame} with sample names, as well as scores for stromal,
#'   immune, and ESTIMATE scores per tumor. If \code{is_affymetrix = TRUE},
#'   purity scores as well.
#' 
#' 
#' Purity scores can be interpreted absolutely: a purity of 0.9 means that tumor
#' is likely 90% pure by this metric. In the case where purity scores are not
#' available (such as in RNAseq), ESTIMATE scores can only be interpreted
#' relatively: a sample that has a lower ESTIMATE score than another in one
#' study can be regarded as more pure than another, but its absolute purity
#' cannot be inferred, nor can purity across other studies be inferred.
#'
#'
#' @references
#'
#' Barbie et al. (2009) <doi:10.1038/nature08460>
#'
#' Yoshihara et al. (2013) <doi:10.1038/ncomms3612>
#'
#'
#' @export
#' @importFrom rlang .data
#' @examples
#' filter_common_genes(ov, id = "hgnc_symbol", tidy = FALSE, tell_missing = TRUE, find_alias = TRUE) |> 
#'   estimate_score(is_affymetrix = TRUE)

estimate_score <- function(df, is_affymetrix) {

    if(missing(is_affymetrix)) {
      stop("`is_affymetrix` must be TRUE or FALSE, and has no default.")
    }
  
    rownames <- df[, 1]
    df <- df[, -1] 
    
    # Sample rank normalization
    df <- df |>  
        apply(2, as.numeric) |> 
        data.matrix()|> 
        apply(2, rank, ties.method = "average")
    
    rownames(df) <- unlist(rownames)
    
    df <- 10000*df/nrow(df)

    gene_set_names <- colnames(tidyestimate::gene_sets)
    scores <- matrix(NA_real_, nrow = ncol(df), ncol = 2)
    
    for (i in 1:2) {
        gene_set <- tidyestimate::gene_sets[, i, drop = FALSE]
        common_genes <- intersect(unlist(gene_set), rownames(df))
        
        message(glue::glue("
                           Number of {colnames(gene_set)} genes in data: \\
                           {length(common_genes)} (out of {nrow(gene_set)})
                           
                           ", .sep = "\n\n"))
        
        if (length(common_genes) == 0) { 
            next
        }
        
        ES.vector <- vector(length = ncol(df))
        
        for (sample in seq_along(1:ncol(df))) {
            
            # Arrange genes by INCREASING expression (remember, ranked previously)
            gene_list <- order(df[, sample], decreasing = TRUE)   
            ordered_genes <- df[gene_list, sample]
            
            hit_ind <- names(ordered_genes) %in% common_genes
            no_hit_ind <- 1 - hit_ind 
            
            ordered_genes <- ordered_genes^0.25
            
            hit_exp <- ordered_genes[hit_ind]
            
            no_hit_penalty <- 
                (no_hit_ind/sum(no_hit_ind)) |> 
                cumsum()
            
            hit_reward <- 
                ((hit_ind*ordered_genes)/sum(hit_exp)) |> 
                cumsum()

            ES.vector[sample] <- sum(hit_reward - no_hit_penalty)
        }
        scores[, i] <- ES.vector
    }

    scores <- scores |> 
        as.data.frame() |> 
        stats::setNames(c("stromal", "immune")) |> 
        dplyr::mutate(sample = colnames(df),
                      estimate = .data$stromal + .data$immune) |> 
        dplyr::relocate(sample)
   
    if (is_affymetrix) {
        # Calculate ESTIMATE-based tumor purity
        scores <- scores |> 
            dplyr::mutate(purity = cos(0.6049872018 + 0.0001467884 * .data$estimate),
                          purity = ifelse(.data$purity < 0, NA, .data$purity))
    }
    
    scores
}


R/filter_common_genes.R
#' Remove non-common genes from data frame
#'
#' As ESTIMATE score calculation is sensitive to the number of genes used, a set
#' of common genes used between six platforms has been established (see
#' \code{?tidyestimate::common_genes}). This function will filter for only those
#' genes.
#'
#' @param df a \code{data.frame} of RNA expression values, with columns corresponding
#'   to samples, and rows corresponding to genes. Either rownames or the first
#'   column can contain gene IDs (see \code{tidy})
#' @param id either \code{"entrezgene_id"} or \code{"hgnc_symbol"}, whichever
#'   \code{df} contains.
#' @param tidy logical. If rownames contain gene identifier, set \code{FALSE}.
#'   If first column contains gene identifier, set \code{TRUE}
#' @param tell_missing logical. If \code{TRUE}, prints message of genes in
#'   common gene set that are not in supplied data frame.
#' @param find_alias logical. If \code{TRUE} and \code{id = "hgnc_symbol"}, will
#'   attempt to find if genes missing from \code{common_genes} are going under
#'   an alias. See details for more information.
#'
#' @details
#'
#' The \code{find_aliases} argument will attempt to find aliases for HGNC
#' symbols in \code{tidyestimate::common_genes} but missing from the provided
#' dataset. This will only run if \code{find_aliases = TRUE} and \code{id =
#' "hgnc_symbol"}.
#'
#' This algorithm is very conservative: It will only make a match if the gene
#' from the common genes has only one alias that matches with only one gene from
#' the provided dataset, \emph{and} the gene from the provided dataset with
#' which it matches only matches with a single gene from the list of common
#' genes. (Note that a single gene may have many aliases). Once a match has been
#' made, the gene in the provided dataset is updated to the gene name in the
#' common gene list.
#'
#' While this method is fairly accurate, is is also a heuristic. Therefore, it is
#' disabled by default. Users should check which genes are becoming reassigned
#' to ensure accuracy.
#'
#' The method of generation of these aliases can be found at
#' \code{?tidyestimate::common_genes}
#'
#' @return A \code{tibble}, with HGNC symbols as the first column
#' @export
#' @importFrom rlang .data
#' @examples
#' filter_common_genes(ov, id = "hgnc_symbol", tidy = FALSE, tell_missing = TRUE, find_alias = FALSE)

filter_common_genes <- function (df, id = c("entrezgene_id", "hgnc_symbol"),
                                 tidy = FALSE, tell_missing = TRUE, 
                                 find_alias = FALSE) {
  if (!tidy) {
    df <- dplyr::as_tibble(df, rownames = id)
  } 
  
  # Assumes first column contains gene IDs
  user_gene_col <- names(df)[1]

  if (sum(duplicated(df[[user_gene_col]])) > 0) warning("Input dataframe has duplicated IDs")
  
  common_genes <- tidyestimate::common_genes
  
  unique_common_genes <- unique(common_genes[,1:2])
  
  filtered <- dplyr::semi_join(df, common_genes, by = stats::setNames(id, user_gene_col)) 
  
  missing <- dplyr::anti_join(common_genes, df, by = stats::setNames(user_gene_col, id))[[user_gene_col]]
  
  if (find_alias & id == "hgnc_symbol") {
    filtered <- find_alias(df, common_genes, missing, filtered)
    missing <- dplyr::anti_join(common_genes, filtered, by = stats::setNames(user_gene_col, "hgnc_symbol"))[["hgnc_symbol"]]
  }
  
  if (tell_missing) tell_missing(missing)
  
  message(glue::glue("
                     
                     Found {nrow(filtered)} of {nrow(unique_common_genes)} \\
                     genes ({round(nrow(filtered)/nrow(unique_common_genes) * 100, 3)}%) \\
                     in your dataset.
                     
                     ", sep = "\n\n"))
  
  if (id == "entrezgene_id") {
    hgnc_entrez <- dplyr::select(common_genes, "hgnc_symbol", "entrezgene_id")
    filtered <- dplyr::left_join(filtered, hgnc_entrez, by = stats::setNames("entrezgene_id", user_gene_col)) |> 
      dplyr::select(-user_gene_col) |> 
      dplyr::relocate("hgnc_symbol")
  }
  
  filtered
}

tell_missing <- function(missing) {
  if (length(unique(missing)) > 10) {
    message(glue::glue("
                       
                       The following genes are in the list of common genes, \\
                       but not in your dataset:
                       
                       {paste(unique(missing)[1:10], collapse = ' ')}...and {length(unique(missing)) - 10} others
                       
                       ",
                       .sep = "\n\n"))
  } else{    
    message(glue::glue("
                       
                       The following genes are in the list of common genes, \\
                       but not in your dataset:
                       
                       {paste(unique(missing), collapse = ' ')}
                       
                       ",
                       .sep = "\n\n"))
  }
}

find_alias <- function(df, common_genes, missing, filtered) {

  df <- dplyr::mutate(df, row_id = rownames(df))

  matched_aliases <- common_genes |>
    dplyr::semi_join(as.data.frame(missing), 
                     by = c("hgnc_symbol" = "missing")) |>
    dplyr::inner_join(df, 
                      by = c("external_synonym" = "hgnc_symbol")) |> 
    dplyr::group_by(.data$hgnc_symbol) |>
    dplyr::mutate(hgnc_has_one_match = dplyr::if_else(dplyr::n() == 1, 
                                                            TRUE, 
                                                            FALSE)) |>
    dplyr::group_by(.data$row_id) |>
    dplyr::mutate(alias_has_one_match = dplyr::if_else(dplyr::n() == 1, 
                                                             TRUE, 
                                                             FALSE)) |>
    dplyr::filter(.data$hgnc_has_one_match & .data$alias_has_one_match) |>
    dplyr::rename(hgnc_symbol_new = .data$hgnc_symbol) |>
    dplyr::select(.data$row_id, .data$hgnc_symbol_new)
  
  df <- df |> 
    dplyr::right_join(matched_aliases, by = "row_id") |> 
    dplyr::mutate(alias = dplyr::if_else(!is.na(.data$hgnc_symbol_new), 
                                         .data$hgnc_symbol_new, 
                                         .data$hgnc_symbol)) |>
    dplyr::select(-.data$hgnc_symbol_new, -.data$hgnc_symbol, -.data$row_id) |>
    dplyr::relocate(.data$alias) |>
    dplyr::rename(hgnc_symbol = .data$alias)
  message(glue::glue("{nrow(df)} of {length(unique(missing))} missing genes found matches using aliases."))
  filtered <- dplyr::bind_rows(filtered, df)
}

R/data.R
#' Genes shared between six expression platforms
#'
#' The ESTIMATE model was trained on a set of genes shared between six
#' expression profiling platforms. Those genes are listed in this dataset.
#'
#' @description As the ESTIMATE model was trained on a specific set of genes,
#'   only those within this dataset should be included before running
#'   \code{estimate_scores}.
#'
#'   These are the genes common to 6 platforms:
#'
#'   - Affymetrix HG-U133Plus2.0
#'
#'   - Affymetrix HT-HG-U133A
#'
#'   - Affymetrix Human X3P
#'
#'   - Agilent 4x44K (G4112F)
#'
#'   - Agilent G4502A
#'
#'   - Illumina HiSeq RNA sequence
#'
#'   The Entrez IDs for the original 10412 genes were matched to HGNC symbols
#'   using \code{biomaRt}. Duplicates and blank entries were filtered. As some
#'   have now been discovered to be pseudogenes or have been deprecated, 22
#'   genes (at time of writing, June 2021) that were in the ESTIMATE package do
#'   not exist here.
#'
#'   As one gene can have multiple synonyms/aliases, and there is only one alias
#'   per line, the number of rows in the data frame (26339) does not reflect the
#'   number of unique genes in the dataset (10391).
#'
#' @format A data frame with 26339 rows and 3 variables: \describe{
#'   \item{entrezgene_id}{Entrez id of the gene} \item{hgnc_symbol}{Human Genome
#'   Organisation (HUGO) Gene Nomenclature Committee symbol}
#'   \item{external_synonym}{A synonym/alias a given gene may go by or
#'   previously went by} }
#' @source
#' \url{https://r-forge.r-project.org/scm/viewvc.php/pkg/estimate/data/common_genes.RData?root=estimate&view=log}
"common_genes"

#' Gene sets to infer tumor stromal and immune infiltration
#'
#' @description Two gene sets, each 141 genes in length, created to infer
#'   stromal and immune infiltration
#'   
#' @format A data frame with 141 row and 2 variables:
#'   \describe{
#'     \item{stromal_signature}{Geneset of HGNC symbols used to infer tumor stromal cell infiltration}
#'     \item{immune_signature}{Geneset of HGNC symbols used to infer tumor immune cell infiltration}
#'   }
#' @source 
#'   \url{https://r-forge.r-project.org/scm/viewvc.php/pkg/estimate/data/SI_geneset.RData?root=estimate&view=log}
"gene_sets"

#' Ovarian cancer tumor RNA expression
#'
#' @description A matrix containing RNA expression of 10 ovarian cancer tumors,
#'   measured using the Affymetrix U133Plus2.0 platform. These data have been
#'   rounded to the 4th decimal place to reduce file size.
#'
#' @format A matrix with 17256 rows and 10 columns, where each column represents
#'   a tumor, and each row represents a gene. Genes are represented by HGNC
#'   symbols in the rownames.
#' @source
#'   \url{https://r-forge.r-project.org/scm/viewvc.php/pkg/estimate/inst/extdata/sample_input.txt?root=estimate&view=log}
"ov"

#' Affymetrix data used to train ESTIMATE algorithm
#'
#' @description A data frame containing the ABSOLUTE-measured and
#'   ESTIMATE-predicted purity values of 995 tumors. Additionally, stromal and
#'   immune scores as calculated by ESTIMATE. All tumors were profiled on
#'   Affymetrix arrays, and were used to generate the Affymetrix algorithm.
#'
#' @format A data frame with 995 rows and 7 variables: 
#'   \describe{
#'     \item{purity_observed}{The purity of a tumor given by ABSOLUTE, ranging from 0 (least pure) to 1 (most pure)} 
#'     \item{stromal}{Stromal infiltration score, as measured by ESTIMATE}
#'     \item{immune}{Immune infiltration score, as measured by ESTIMATE}
#'     \item{estimate}{ESTIMATE score, calculated by the sum of immune and stromal scores}
#'     \item{purity_predicted}{Tumor purity inferred using the ESTIMATE algorithm}
#'     \item{ci_95_low}{Lower bound of a 95\% confidence interval of predicted purity scores}
#'     \item{ci_95_high}{Upper bound of a 95\% confidence interval of predicted purity scores}
#'   }
#'   
#' @source 
#'   \url{https://r-forge.r-project.org/scm/viewvc.php/pkg/estimate/data/PurityDataAffy.RData?root=estimate&view=log}
"purity_data_affy"
